pkg/trie/triedb/triedb_iterator.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+
+	nibbles "github.com/ChainSafe/gossamer/pkg/trie/codec"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+type iteratorState struct {
+	parentFullKey []byte     // key of the parent node of the actual node
+	node          codec.Node // actual node
+}
+
+// fullKeyNibbles return the full key of the node contained in this state
+// child is the child where the node is stored in the parent node
+func (s *iteratorState) fullKeyNibbles(child *int) []byte {
+	fullKey := bytes.Join([][]byte{s.parentFullKey, s.node.GetPartialKey()}, nil)
+	if child != nil {
+		return bytes.Join([][]byte{fullKey, {byte(*child)}}, nil)
+	}
+
+	return nibbles.NibblesToKeyLE(fullKey)
+}
+
+type TrieDBIterator struct {
+	db        *TrieDB          // trie to iterate over
+	nodeStack []*iteratorState // Pending nodes to visit
+}
+
+func NewTrieDBIterator(trie *TrieDB) *TrieDBIterator {
+	rootNode, err := trie.getRootNode()
+	if err != nil {
+		panic("trying to create trie iterator with incomplete trie DB")
+	}
+	return &TrieDBIterator{
+		db: trie,
+		nodeStack: []*iteratorState{
+			{
+				node: rootNode,
+			},
+		},
+	}
+}
+
+// nextToVisit sets the next node to visit in the iterator
+func (i *TrieDBIterator) nextToVisit(parentKey []byte, node codec.Node) {
+	i.nodeStack = append(i.nodeStack, &iteratorState{
+		parentFullKey: parentKey,
+		node:          node,
+	})
+}
+
+// nextState pops the next node to visit from the stack
+func (i *TrieDBIterator) nextState() *iteratorState {
+	currentState := i.nodeStack[len(i.nodeStack)-1]
```

Comment:
can we check if `i.nodeStack` is not empty? or include a comment like `warn: this function does not check if the node stack is empty, this check is made at NextEntry method`

Reply:
We are checking it [here](https://github.com/ChainSafe/gossamer/blob/d80fdae4f2e1b99f9a755e3a6469a37100abb756/pkg/trie/triedb/triedb_iterator.go#L65) that's the only place where the `nextState()` is call.
But I'll add the comment to let know that the caller should check it 👍